File: macbook/bettertouchtool/repair_rps.js
// Worker for repair_rps.py. Only creates the missing child action; never runs it.
ObjC.import('Foundation');
const RPS_MEDIA = 'D9B0ED12-C4BE-4E74-B0DA-0CC3BE092289';
const RPS_BUTTON = '103D7824-47C1-4B1B-9106-71E4997BCB58';
const RPS_ACTION = '525212E9-F6F3-4CF0-A9F7-86C1A52EEB30';
const RPS_SYNC = 'DA200AE7-BB01-4AB5-89A7-CF1587728D7D';

function rpsAction() {
    return {BTTUUID:RPS_ACTION,BTTTriggerParentUUID:RPS_BUTTON,
        BTTTriggerType:-1,BTTTriggerClass:'BTTTriggerTypeFloatingMenu',
        BTTEnabled:1,BTTEnabled2:1,BTTOrder:0,BTTActionCategory:0,
        BTTPredefinedActionType:248,
        BTTPredefinedActionName:'Trigger Named Trigger (Configured in Other Tab)',
        BTTNamedTriggerToTrigger:'Sync RPi Scripts'};
}
function rpsButton(media) {
    if (!media || media.BTTUUID!==RPS_MEDIA || Number(media.BTTTriggerType)!==767 ||
        !Array.isArray(media.BTTMenuItems)) throw new Error('Cannot export complete Media menu.');
    const matches=media.BTTMenuItems.filter(x=>x.BTTUUID===RPS_BUTTON);
    if(matches.length!==1 || Number(matches[0].BTTTriggerType)!==773)throw new Error('RPS button missing or changed.');
    return matches[0];
}
function rpsApply(btt, snapshot, fingerprint) {
    function get(id) {
        const raw=JSON.parse(btt.get_trigger(id));
        const hits=(Array.isArray(raw)?raw:[raw]).filter(x=>x&&x.BTTUUID===id);
        if(hits.length!==1)throw new Error('Cannot export '+id);
        return hits[0];
    }
    const before=get(RPS_MEDIA),sync=get(RPS_SYNC),button=rpsButton(before);
    if(fingerprint(before)!==fingerprint(snapshot.media) || fingerprint(sync)!==fingerprint(snapshot.sync)) {
        throw new Error('BTT changed during backup; no changes made.');
    }
    if((button.BTTMenuItemActions||[]).length || (button.BTTAdditionalActions||[]).length) {
        throw new Error('RPS already has runtime actions; refusing to duplicate or replace them.');
    }
    btt.add_new_trigger(JSON.stringify(rpsAction()),{parent_uuid:RPS_BUTTON});
    const after=get(RPS_MEDIA),restored=rpsButton(after),actions=restored.BTTMenuItemActions||[];
    if(actions.length!==1 || actions[0].BTTUUID!==RPS_ACTION ||
        Number(actions[0].BTTPredefinedActionType)!==248 ||
        actions[0].BTTNamedTriggerToTrigger!=='Sync RPi Scripts' ||
        Number(actions[0].BTTActionCategory||0)!==0 || Number(actions[0].BTTOrder||0)!==0 ||
        Number(actions[0].BTTEnabled===undefined?1:actions[0].BTTEnabled)!==1) {
        throw new Error('Restored RPS action failed API readback.');
    }
    // Ignore only the deliberately added action when checking the existing tree.
    if(Object.prototype.hasOwnProperty.call(button,'BTTMenuItemActions'))restored.BTTMenuItemActions=[];
    else delete restored.BTTMenuItemActions;
    if(fingerprint(after)!==fingerprint(before) || fingerprint(get(RPS_SYNC))!==fingerprint(sync)) {
        throw new Error('Unexpected change outside the restored RPS action.');
    }
    return 'Created and verified the RPS action. No sync executed.';
}
function run(argv) {
    const mode=argv[0],repo=argv[1];
    function read(path) {
        const text=$.NSString.stringWithContentsOfFileEncodingError(path,$.NSUTF8StringEncoding,null);
        if(!text)throw new Error('Cannot read '+path);
        return ObjC.unwrap(text);
    }
    const btt=Application('/Applications/BetterTouchTool.app');
    if(mode==='snapshot') {
        function get(id) {
            const raw=JSON.parse(btt.get_trigger(id));
            const hits=(Array.isArray(raw)?raw:[raw]).filter(x=>x&&x.BTTUUID===id);
            if(hits.length!==1)throw new Error('Cannot export '+id);
            return hits[0];
        }
        const media=get(RPS_MEDIA);rpsButton(media);
        return JSON.stringify({media,sync:get(RPS_SYNC)});
    }
    if(mode!=='apply')throw new Error('Unknown worker mode.');
    const fingerprint=new Function(read(repo+'/macbook/bettertouchtool/btt_common.js')+
        '\nreturn BTTCommon.fingerprint;')();
    return rpsApply(btt,JSON.parse(read(argv[2])),fingerprint);
}

File: macbook/bettertouchtool/repair_rps.py
#!/usr/bin/env python3
"""Restore only RPS's missing named-trigger action; never execute a deployment."""
from __future__ import annotations

import argparse
from contextlib import closing
import json
import os
from pathlib import Path
import sqlite3
import subprocess
import sys
import time

from btt_common import ROOT, MEDIA, connect, current_database, records, check_entity, backup_configuration

BUTTON = '103D7824-47C1-4B1B-9106-71E4997BCB58'
ACTION = '525212E9-F6F3-4CF0-A9F7-86C1A52EEB30'
SYNC = 'DA200AE7-BB01-4AB5-89A7-CF1587728D7D'
NAME = 'Sync RPi Scripts'
COMMAND = str(ROOT / 'pi/sync_scripts.sh')
WORKER = Path(__file__).with_suffix('.js')
ENTITY = {'uuid': ACTION, 'parent': BUTTON, 'node': {
    'BTTUUID': ACTION, 'BTTTriggerType': -1, 'BTTOrder': 0,
    'BTTPredefinedActionType': 248, 'BTTNamedTriggerToTrigger': NAME}}


def preflight(database):
    saved = records(database)
    main, button, sync = (saved.get(uid) for uid in (MEDIA, BUTTON, SYNC))
    if not main or not button or button['parent'] != MEDIA or button['ZGESTURETYPE'] != 773:
        raise RuntimeError('Expected RPS button is missing or reparented; no changes made.')
    preset = main['ZBELONGSTOPRESET2']
    if button['ZBELONGSTOPRESET2'] != preset or button['ZACTION'] != 366:
        raise RuntimeError('RPS parent action/preset changed; no changes made.')
    if not sync or sync['ZBELONGSTOPRESET2'] != preset or sync['ZGESTURETYPE'] != 643 or \
            sync['ZACTION'] != 206 or sync['ZLAUNCHPATH'] != COMMAND:
        raise RuntimeError('Current sync trigger no longer points to the expected script; no changes made.')
    with closing(connect(database)) as connection:
        eligible = list(connection.execute('''SELECT t.ZUNIQUEIDENTIFIER
            FROM ZBTTBASEENTITY t JOIN ZBTTBASEENTITY p ON t.ZBELONGSTOPRESET2=p.Z_PK
            WHERE t.ZGESTURETYPE=643 AND t.ZGESTURECONFIG=? AND p.ZACTIVATED<>0
              AND COALESCE(t.ZISENABLED,1)<>0 AND COALESCE(t.ZENABLEDNEW,1)<>0''', (NAME,)))
        if [r[0] for r in eligible] != [SYNC]:
            raise RuntimeError('Sync named-trigger resolution is disabled or ambiguous; no changes made.')
        disabled = connection.execute('''SELECT count(*) FROM ZBTTBASEENTITY
            WHERE ZUNIQUEIDENTIFIER IN (?,?) AND (ZISENABLED=0 OR ZENABLEDNEW=0)''',
            (MEDIA, BUTTON)).fetchone()[0]
        if disabled:
            raise RuntimeError('Media or RPS is disabled; no changes made.')
    children = [r for r in saved.values() if r['parent'] == BUTTON]
    action = saved.get(ACTION)
    if action:
        error = check_entity(action, ENTITY, preset)
        if error or len(children) != 1:
            raise RuntimeError('Existing RPS action differs from the expected link; refusing to overwrite it.')
        with closing(connect(database)) as connection:
            row = connection.execute('SELECT ZISENABLED,ZENABLEDNEW FROM ZBTTBASEENTITY WHERE ZUNIQUEIDENTIFIER=?',
                                     (ACTION,)).fetchone()
            if any(value == 0 for value in row):
                raise RuntimeError('Existing RPS action is disabled; refusing to replace it.')
        return True
    if children:
        raise RuntimeError('RPS has other saved actions; refusing to overwrite them.')
    return False


def worker(mode, argument=''):
    result = subprocess.run(['/usr/bin/osascript', '-l', 'JavaScript', str(WORKER), mode, str(ROOT), str(argument)],
                            capture_output=True, text=True, timeout=60)
    if result.returncode:
        raise RuntimeError(result.stderr.strip() or 'BTT API request failed.')
    return result.stdout.strip()




def repair(inspect=False):
    os.umask(0o077)
    database = current_database()
    installed = preflight(database)
    if inspect:
        print(json.dumps({'rps_link_saved': installed, 'sync_script': COMMAND, 'will_execute_sync': False}))
        return
    if installed:
        print('RPS link is already saved correctly. No changes or sync performed.')
        return
    if not Path(COMMAND).is_file():
        raise RuntimeError('Sync script is missing; no changes made.')
    snapshot = json.loads(worker('snapshot'))
    backup = backup_configuration(database, snapshot, prefix='btt-rps-backup-')
    print('Verified BTT configuration snapshot: '+str(backup.parent), flush=True)
    if preflight(database):
        print('RPS was repaired while preparing the backup; no changes or sync performed.')
        return
    try:
        print(worker('apply', backup), flush=True)
        deadline = time.monotonic()+5
        while not preflight(database):
            if time.monotonic() >= deadline:
                raise RuntimeError('BTT has not saved the action to disk; repair is not verified.')
            time.sleep(.1)
    except (RuntimeError, subprocess.SubprocessError) as error:
        raise RuntimeError(str(error)+'\nBackup: '+str(backup.parent)) from error
    print('Verified saved RPS -> Sync RPi Scripts link. No deployment or BTT restart performed.')


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('--inspect', action='store_true', help='read-only saved-configuration check')
    parser.add_argument('--with-escape', action='store_true', help='also install visible-dropdown-only Escape dismissal')
    args = parser.parse_args()
    repair(args.inspect)
    if args.with_escape:
        from install_menu_escape import install
        install(args.inspect)


if __name__ == '__main__':
    try:
        main()
    except (RuntimeError, OSError, ValueError, sqlite3.Error, subprocess.SubprocessError) as error:
        print(str(error), file=sys.stderr)
        raise SystemExit(1)
